### DISCUSSION TITLE
Faststream application does not close it in case of error on startup event!

### DISCUSSION THREAD

**Discussion by fraFerrari99** [ag2ai/faststream#1910] [Q&A]
Hi, 
I'm new with Faststream but I'm facing a particular error, that is related to the fact that when, inside the on_startup event, the application faces some errors, it calls the on_shutdown event but does not really close, it remains open! I can add the sys.exit(-1) line of code in the on_shutdown method, but I wanted to ask you why it does not close it at the end of the on_shutdown?

**Comment by Lancetnik** [comment]
> Hi, I'm new with Faststream but I'm facing a particular error, that is related to the fact that when, inside the on_startup event, the application faces some errors, it calls the on_shutdown event but does not really close, it remains open! I can add the sys.exit(-1) line of code in the on_shutdown method, but I wanted to ask you why it does not close it at the end of the on_shutdown?

Can you show me the code, please?

We following snippet works as expected - raises the error to console at startup and downs

```python
from faststream import FastStream
from faststream.rabbit import RabbitBroker

broker = RabbitBroker()
app = FastStream(broker)

@app.on_startup
async def _():
    raise ValueError
```

**Reply by fraFerrari99** [reply]
These are the `on_startup` and `on_shutdown` methods of the faststream application.
```python
@app.on_startup
async def on_startup():
    try:
        global sem
        init_logging()

        try:
            await broker.connect()
        except Exception as e:
            logger.error(f"Failed to connect to the broker: {str(e)}")
            sys.exit(-1)

        if launch_executors_on_startup:
            await do_launch_executors_on_startup()
        executors = await initialize_executors_client()
        logger.info(
            "Application can now receive messages, with parallelism %s", len(executors)
        )
    except Exception as e:
        logger.error(f"Failed to start application: {str(e)}")
        sys.exit(-1)

@app.on_shutdown
async def on_shutdown():
    logger.info("Received shutdown signal")
    await terminate_executors()
    sys.exit(-1)
```

To prevent the application from remaining active after errors occur during startup, I've added two `try/except` blocks within `on_startup`. The first handles broker connection errors specifically, ensuring the application exits cleanly if the connection fails. The second, wrapping the entire method, catches any other initialization errors.

The `sys.exit(-1)` in `on_shutdown` ensures that the application terminates fully, addressing cases where issues arise after startup and prevent the app from closing properly. While I initially thought these issues might be related to running with Uvicorn and the `as_asgi` method, integrating `on_startup` and `on_shutdown` events into the FastStream application and then adding the ASGI compatibility didn't solve the problem. 

So the main way to solve these issues is to use the `after_startup`?

**Reply by Lancetnik** [reply]
**FastStream** shouldn't start if any error in `@app.on_startup` occures - I showed you the code example. Which broker are you using? Also, did you try lifespan instaed of hooks? - it is much suitable for your case

**Reply by fraFerrari99** [reply]
Thanks for getting back to me!

I'm using RabbitBroker, and I had to move the connection test into the `on_startup` method. I noticed that if there’s a connection error—like an incorrect port—it throws an error but doesn’t actually terminate the application. I also tried managing it with the `lifespan`, but if there’s an issue during the `yield` phase, it still doesn't close the app as expected.

**Reply by Lancetnik** [reply]
I still can't reproduce your problem, sorry

My testing code snippet

```python
from faststream import FastStream
from faststream.rabbit import RabbitBroker

broker = RabbitBroker()
app = FastStream(broker)

@app.on_startup
async def t():
    await broker.connect(port=8888)
```

Raises the following error to console

```console
AMQPConnectionError: Multiple exceptions: [Errno 61] Connect call failed ('::1', 8888, 0, 0), [Errno 61] Connect call failed ('127.0.0.1', 8888)
```

**Reply by Lancetnik** [reply]
I think, your statup problem can be related not to **FastStream** itself, but to your tasks you running in the hook

**Reply by fraFerrari99** [reply]
With the connection part inside of the startup, if there's an error, it closes the application like excepted but the problem is that, in case of any errors found after the startup, it doesn't close the application but it just logs the error and the application remains alive while I would like to close it.

I tried, like you said before, with the lifespan but the situation didn't change. There's some way to handle this possible situation?

**Reply by fraFerrari99** [reply]
You were speaking about the after_startup method, but I didn't clearly understand how to use it to handle the various possible errors.

**Reply by Lancetnik** [reply]
@fraFerrari99 `after_startup` was just a mistake, sorry. I mean `on_startup`

**Reply by Lancetnik** [reply]
> in case of any errors found after the startup, it doesn't close the application but it just logs the error and the application remains alive while I would like to close it.

**FastStream** was designed to wait for connection recovery endless time to allow you don't care about infrastructure availability. If you connected at startup and when connection broken, service just waiting for recovery.

**Reply by fraFerrari99** [reply]
> > in case of any errors found after the startup, it doesn't close the application but it just logs the error and the application remains alive while I would like to close it.
> 
> **FastStream** was designed to wait for connection recovery endless time to allow you don't care about infrastructure availability. If you connected at startup and when connection broken, service just waiting for recovery.

Our issue was exactly this: the FastStream application was starting before RabbitMQ, and as a result, it was unable to establish a connection. However, the problem persisted even after RabbitMQ was up and running. For this reason we put that code with `broker.connect()` inside the on_startup in the try/except block with `sys.exit(-1)` in case of error, because in this way if there's an error and the pod on faststream application is down, K8 will restart it and maybe in that case it will be able to estabilish the connection with RabbitMQ.

**Reply by Sehat1137** [reply]
@fraFerrari99 
Pls, give me MRE of ur code, I have a few ideas why this hapend, but i need check the code
